torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1152,6 +1169,279 @@ def call_function(
         )
 
 
+class CannotConvertRangeToHigherOrder(TorchDynamoException):
+    pass
+
+
+class RangeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    The general idea is that for a simple for loop like
+
+    ```
+    for i in range(10):
+        x = x + 1
+    ```
+
+    We can convert the loop body to a function and the loop itself
+    to a series of calls:
+
+    ```
+    def loop_body(local_0, local_1, local_...):
+        local_1 = local_1 + 1
+        return (local_0, local_1, local_...)
+
+    (x, i) = loop_body(x, i)
+    (x, i) = loop_body(x, i)
+    (x, i) = loop_body(x, i)
+    # ... 7 more times
+    ```
+
+    There are a few benefits to doing this, but mainly we bypass a ton
+    of stuff associated with tracing for loops. That itself provides some speedup.
+
+    Once we implement function compilation caching, or a higher order operator,
+    like jax.lax.fori_loop, for loops will compile in constant time, and perhaps even
+    better than what we currently have.
+    """
+
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
+        "RETURN_VALUE",
+        "YIELD_VALUE",
+        "YIELD_FROM",
+        "GET_YIELD_FROM_ITER",
+        "SEND",
+        "MAKE_FUNCTION",
+        "LOAD_DEREF",  # This could be supported in the future.
+        # Exceptions
+        "RAISE_VARARGS",
+        "SETUP_FINALLY",
+        "POP_EXCEPT",
+        # These two are fine for the end of the loop body but not within.
+        "JUMP_BACKWARD",
+        "JUMP_ABSOLUTE",
+    }
+
+    store_target: int
+
+    def __init__(
+        self, value: "RangeIteratorVariable", source: Optional[Source] = None, **kwargs
+    ):
+        from ..source import GlobalSource
+
+        super().__init__(value, source, **kwargs)
+        self.store_target = -1
+        self.source = GlobalSource("range")
```

Comment:
It's weird to take an optional source arg, pass it to super, and then blow out the self.source like this. I would just reject and assert that no source is ever passed here.